### PR TITLE
Redirect incoming information for deleted records to the deduped record

### PR DESCRIPTION
**Pull request by ssrihari**
**Story card:** [ch2682](https://app.clubhouse.io/simpledotorg/story/2682/handle-incoming-information-from-sync-on-a-merged-and-soft-deleted-patient)

## Because

After patients have been deduplicated, any new information or updates to the deleted patients, or their associated records should go to the deduped records.

## This addresses

Redirection of the incoming request on deleted records to the deduplicated records
- Needs integration tests

### CODE REVIEW

File: app/transformers/api/v3/transformer.rb [hunk 1]
```diff
@@ -1,7 +1,19 @@
...
+      return attributes unless deduped_record || deduped_patient
+
+      attributes["id"] = deduped_record.id if deduped_record.present?
+      attributes["patient_id"] = deduped_patient.id if deduped_patient.present? && attributes["patient_id"].present?
+      attributes
+    end
+
     def from_request(payload_attributes)
       rename_attributes(payload_attributes, from_request_key_mapping)
+        .then { |attributes| redirect_to_deduped_patient(attributes) }
```
Comment (harimohanraj89) [inline app/transformers/api/v3/transformer.rb:16]: Oof, I'm loving the addition of `then` to Ruby.

File: app/transformers/api/v3/transformer.rb [hunk 2]
```diff
@@ -1,7 +1,19 @@
 class Api::V3::Transformer
   class << self
+    def redirect_to_deduped_patient(attributes)
+      # NOTE: Move this to a different layer if/when this becomes more complex
+      deduped_record = DeduplicationLog.find_by_deleted_record_id(attributes["id"])&.deduped_record
```
Comment (harimohanraj89) [inline app/transformers/api/v3/transformer.rb:5]: I think it's encouraged to move away from the magic `find_by_something` methods, to the less-meta-programmy `find_by(attribute: value)` format. Shall we do that?

File: app/transformers/api/v3/transformer.rb [hunk 3]
```diff
@@ -1,7 +1,19 @@
...
+      deduped_patient = DeduplicationLog.find_by_deleted_record_id(attributes["patient_id"])&.deduped_record
```
Comment (harimohanraj89) [inline app/transformers/api/v3/transformer.rb:6]: How does this clause work when the record itself is a patient? Does it behave gracefully?

**Review by harimohanraj89 — APPROVED** [review]
Tiny thoughts. Looks great.

File: spec/controllers/shared_examples/shared_spec_for_sync_controller.rb [hunk 4]
```diff
@@ -180,6 +180,24 @@ def discard_patient(record)
                            .with_int_timestamps)
       end
     end
+
+    it "updates deduped records if a record was merged" do
+      deduped_record = create_record_list(1).first
+      merged_record = existing_records.first
+      updated_record = updated_records.first
```
Comment (ssrihari) [inline spec/controllers/shared_examples/shared_spec_for_sync_controller.rb:187]: s/merge/delete

**Pull request merged**